---
title: "Exploration of CD4 ChIP-Seq Peak Sizes"
author: "Ryan C. Thompson"
date: '`r gsub("\\s+", " ", format(Sys.time(), "%B %e, %Y"))`'
output:
  html_document: default
  html_notebook: default
---

# Preliminary Setup

```{r setup, include=FALSE, cache=FALSE}
knitr::opts_chunk$set(
    echo = TRUE, retina=2, cache=TRUE, autodep=TRUE,
    cache.extra = list(params=params), 
    fig.height=8, fig.width=8,
    cache.path = paste0(here::here("cache", "chipseq-peak-sizes"),
                        .Platform$file.sep)
```

First we load the necessary libraries, along with a set of utility functions.

```{r load_packages, message=FALSE, cache=FALSE}
library(here)
library(stringr)
library(magrittr)
library(dplyr)
library(SummarizedExperiment)
library(ggplot2)
library(assertthat)
library(RColorBrewer)
library(modeest)

library(BSgenome.Hsapiens.UCSC.hg38)

library(doParallel)
ncores <- getOption("mc.cores", default=parallel::detectCores(logical = FALSE))
options(mc.cores=ncores)
registerDoParallel(cores=ncores)
library(BiocParallel)
register(DoparParam())

options(future.globals.maxSize=4 * 1024^3)
library(future)
plan(multicore)

source(here("scripts/utilities.R"))
```

# Data Loading and Preprocessing

First we load the consensus peaks called from the reads pooled from all samples.

```{r load_peaks}
histone.marks <- c("H3K4me3", "H3K4me2", "H3K27me3") %>% setNames(., .)
peakfiles <- here(
    "peak_calls", "epic_hg38.analysisSet",
    str_c(histone.marks, "_condition.ALL_donor.ALL"),
    "peaks_noBL_IDR.narrowPeak") %>% setNames(histone.marks)
assert_that(all(file.exists(peakfiles)))
allpeaks <- bplapply(peakfiles, . %>% read.narrowPeak %>% as("GRanges") %>%
        assign_into(seqinfo(.), seqinfo(BSgenome.Hsapiens.UCSC.hg38)[seqlevels(.)])) %>%
        setNames(.$name)
```

# Peak width analysis

First, we look at the granularity of the peak caller in terms of what peak widths it calls, and what positions it calls them at.

```{r peak_width_granularity}
combined.peaklist <- unlist(GRangesList(allpeaks))
widths <- width(combined.peaklist) %>% table %>% as.data.frame %>% 
    set_colnames(c("Width", "Freq")) %>%
    mutate(Width=as.integer(as.character(Width))) %>%
    arrange(Width)
print(widths)
widths$Width %>% diff %>% table %>% as.data.frame %>% 
    set_colnames(c("Diff", "Freq")) %>%
    mutate(Diff=as.integer(as.character(Diff))) %>%
    arrange(Diff)
```

We can see that all peaks are called with a width that is a multiple of 200 bp, and no peaks smaller than 800 bp are called. This is expected based on the parameters used in the peak caller. We can also look at the distribution of start positions:

```{r peak_start_granularity}
start(combined.peaklist) %>% sort %>% diff %>% table %>% as.data.frame %>% 
    set_colnames(c("Diff", "Freq")) %>%
    mutate(Diff=as.integer(as.character(Diff))) %>%
    arrange(Diff)
```

It appears that the start positions of all peaks are also multiples of 200 bp from each other. Hence, we will use 200 as the bin size for all the histograms below.

## Peak Filtering

We begin by selecting only peaks  with an IDR value of 0.2 or less, then converting them to a data frame in preparation for plotting.

```{r initial_filter_peaks}
idr.threshold <- 0.2
genome.size <- seqlengths(seqinfo(allpeaks[[1]])) %>% as.numeric %>% sum
peaks <- lapply(allpeaks, . %>% .[.$qValue >= -log10(idr.threshold)])
peaktable <- lapply(names(peaks), function(i) cbind(HistoneMark=i, as.data.frame(peaks[[i]]))) %>% 
    do.call(what=rbind)
```

## Peak width histograms

Now we examine the distribution of peak widths for each histone mark. We zoom in on the lower end of the distribution to get more detail there.

```{r peak_width_dist}
q95 <- quantile(peaktable$width, .95)
p <- ggplot(peaktable) + 
    aes(x=width) + 
    geom_histogram(binwidth=200, boundary=100) + 
    facet_wrap(~HistoneMark, ncol=1) + 
    ggtitle("Histogram of peak widths")
ggprint(p + labs(subtitle="Full distribution"))
ggprint(p + coord_cartesian(xlim=c(0, q95)) +
            labs(subtitle="Bottom 95%"))
ggprint(p + coord_cartesian(xlim=c(0, 10000)) + 
            labs(subtitle="Truncated at 10000bp"))
ggprint(p + coord_cartesian(xlim=c(0, 5000)) + 
            labs(subtitle="Truncated at 5000bp"))
```

```{r peak_width_dist_summary}
peaktable %>% group_by(HistoneMark) %>%
    summarize(Min=min(width), Q25=quantile(width, .25), Mode=mfv(width), Median=median(width), Mean=mean(width), Q75=quantile(width, .75), Max=max(width))
```
Based on these distributions, it appears that a "typical" peak width for H3K4me3 and H3K4me2 is between 1500 and 2500 bp, with a long tail of much longer peaks. For H3K27me3, the typical peak size is much larger, with an even larger and fatter tail on the distribution. Based on this, the proposed window width of 500 bp is appropriate for capturing potential within-peak variation, and there is no need to use a smaller window width to capture the bahavior of individual histones within each peak.
